pyg_lib/csrc/ops/cpu/matmul_kernel.cpp

```diff
@@ -1,27 +1,274 @@
+          ld_src0[group_idx] = k;
+          ld_src1[group_idx] = n;
+          ld_dst[group_idx] = n;
+          group_sizes[group_idx] = indices.size();
+          ++group_idx;
+
+          for (const auto tensor_idx : indices) {
+            src0.push_back(input[tensor_idx].data_ptr<scalar_t>());
+            src1.push_back(other[tensor_idx].data_ptr<scalar_t>());
+            dst.push_back(out[tensor_idx].data_ptr<scalar_t>());
+          }
+        }
+
+        auto src0_ptrs = const_cast<const scalar_t**>(src0.data());
+        auto src1_ptrs = const_cast<const scalar_t**>(src1.data());
+        auto dst_ptrs = dst.data();
+
+#if AT_MKL_SEQUENTIAL()
+        // unlikely to happen - requires Torch to be built from source with
+        // explicit flag denoting MKL sequential version
+        parallel_mkl_blas_gemm_batched(src0_ptrs, src1_ptrs, dst_ptrs, ms, ns,
+                                       ks, ld_src0, ld_src1, ld_dst,
+                                       group_sizes, alpha, beta, group_count);
+#else
+        mkl_blas_gemm_batched(ms.data(), ns.data(), ks.data(), alpha.data(),
+                              src0_ptrs, ld_src0.data(), src1_ptrs,
+                              ld_src1.data(), beta.data(), dst_ptrs,
+                              ld_dst.data(), group_count, group_sizes.data());
+#endif
+      });
+}
+
+void grouped_matmul_out_kernel_at_impl(const std::vector<at::Tensor> input,
+                                       const std::vector<at::Tensor> other,
+                                       std::vector<at::Tensor> out) {
+  for (size_t i = 0; i < out.size(); ++i) {
     at::matmul_out(const_cast<at::Tensor&>(out[i]), input[i], other[i]);
+  }
+}
+
+void grouped_matmul_out_kernel(const std::vector<at::Tensor> input,
+                               const std::vector<at::Tensor> other,
+                               std::vector<at::Tensor> out) {
+  AT_DISPATCH_ALL_TYPES(
+      input.front().scalar_type(), "grouped_matmul_out_kernel", [&] {
+        c10::guts::if_constexpr<WITH_MKL_BLAS() && AT_MKL_ENABLED() &&
+                                is_blas_library_type<scalar_t>::value>(
+            [&](auto _) {
+              grouped_matmul_out_kernel_mkl_impl(input, other, out);
+            },
+            [&](auto _) {
+              grouped_matmul_out_kernel_at_impl(input, other, out);
+            });
+      });
 }
 
 std::vector<at::Tensor> grouped_matmul_kernel(const at::TensorList input,
                                               const at::TensorList other) {
-  std::vector<at::Tensor> out(input.size());
-  for (size_t i = 0; i < input.size(); ++i)
-    out[i] = input[i].new_empty({input[i].size(0), other[i].size(-1)});
+  std::vector<at::Tensor> input_contig;
+  std::vector<at::Tensor> other_contig;
+  std::vector<at::Tensor> out;
+
+  for (size_t i = 0; i < input.size(); ++i) {
+    input_contig.push_back(input[i].contiguous());
+    other_contig.push_back(other[i].contiguous());
```

Comment:
use `embrace_back` and `reserve` for std::vectors is usually a good practice, otherwise if the list if long, you will get a lot of memory allocation in your vtune profile.

Reply:
Done.